livy/src/main/java/org/apache/zeppelin/livy/LivyHelper.java

```diff
@@ -0,0 +1,364 @@
+      context.out.clear();
+      Code r = null;
+      String incomplete = "";
+
+      for (int l = 0; l < linesToRun.length; l++) {
+        String s = linesToRun[l];
+        // check if next line starts with "." (but not ".." or "./") it is treated as an invocation
+        //for spark
+        if (l + 1 < linesToRun.length) {
+          String nextLine = linesToRun[l + 1].trim();
+          if (nextLine.startsWith(".")
+              && !nextLine.startsWith("..")
+              && !nextLine.startsWith("./")) {
+            incomplete += s + "\n";
+            continue;
+          }
+        }
+
+        InterpreterResult res;
+        try {
+          res = interpret(incomplete + s, context, userSessionMap);
+        } catch (Exception e) {
+          LOGGER.info("Interpreter exception", e);
+          return new InterpreterResult(Code.ERROR, InterpreterUtils.getMostRelevantMessage(e));
+        }
+
+        r = getResultCode(res);
+
+        if (r == Code.ERROR) {
+          out.setInterpreterOutput(null);
+          return new InterpreterResult(r, "");
+        } else if (r == Code.INCOMPLETE) {
+          incomplete += s + "\n";
+        } else {
+          out.write((res.message() + "\n").getBytes(Charset.forName("UTF-8")));
+          incomplete = "";
+        }
+      }
+
+      if (r == Code.INCOMPLETE) {
+        out.setInterpreterOutput(null);
+        return new InterpreterResult(r, "Incomplete expression");
+      } else {
+        out.setInterpreterOutput(null);
+        return new InterpreterResult(Code.SUCCESS);
+      }
+
+    } catch (Exception e) {
+      LOGGER.error("error in interpretInput", e);
+      return new InterpreterResult(Code.ERROR, e.getMessage());
+    }
+  }
+
+  private Code getResultCode(InterpreterResult r) {
+    if (r.code().equals(Code.SUCCESS)) {
+      return Code.SUCCESS;
+    } else if (r.code().equals(Code.INCOMPLETE)) {
+      return Code.INCOMPLETE;
+    } else {
+      return Code.ERROR;
+    }
+  }
+
+  public InterpreterResult interpret(String stringLines,
+                                        final InterpreterContext context,
+                                        final Map<String, Integer> userSessionMap)
+      throws Exception {
+    stringLines = stringLines
+        //for "\n" present in string
+        .replaceAll("\\\\n", "\\\\\\\\n")
+        //for new line present in string
+        .replaceAll("\\n", "\\\\n")
+        // for \" present in string
+        .replaceAll("\\\\\"", "\\\\\\\\\"")
+        // for " present in string
+        .replaceAll("\"", "\\\\\"");
+
+    if (stringLines.trim().equals("")) {
+      return new InterpreterResult(Code.SUCCESS, "");
+    }
+    Map jsonMap = executeCommand(stringLines, context, userSessionMap);
+    Integer id = ((Double) jsonMap.get("id")).intValue();
+    InterpreterResult res = getResultFromMap(jsonMap);
+    if (res != null) {
+      return res;
+    }
+
+    while (true) {
+      Thread.sleep(1000);
```

Comment:
I thought about timeout here as well, but there can be long running commands like someone may want to download GBs of file which can take hours to download.